NXDOC-2956: document Audit Router and Blue/Green Audit migration (#2715)
nuxeo/doc.nuxeo.com-platform-spaces: src/nxdoc/nuxeo-server/data-store/data-modeling/audit/copy-audit-backend.md

---
title: Copy an Audit Backend
description: Copy log entries from one Audit Backend to another using the audit scroll, the copyAudit bulk action and the Management REST API.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - blue-green-migration
toc: true
tree_item_index: 780
---

{{#> callout type='warning' heading='Available since 2025.19'}}
The audit scroller, the `copyAudit` bulk action and the related Management
REST API endpoints are introduced in Nuxeo 2025.19.
{{/callout}}

{{! excerpt}}
Copying log entries between two Audit Backends is the cornerstone of a
**Blue/Green Audit migration** — for instance when upgrading from OpenSearch 1.x

> to OpenSearch 2.x, or when moving the audit out of SQL Database. Nuxeo provides a

Reviewer: ```suggestion
to OpenSearch 2.x, or when moving the audit out of SQL backend. Nuxeo provides a
```

generic scroller, a `copyAudit` Bulk Action and Management REST endpoints to
trigger and verify the copy.
{{! /excerpt}}

## Audit Scroll

The `audit` scroll is a [Bulk Action Framework]({{page page='bulk-action-framework'}})
scroller backed by the Audit `QueryBuilder` API. It scrolls log entry

> identifiers from one [Audit Backend]({{page page='audit'}}#audit-backends)

Reviewer: ```suggestion
identifiers from one [Audit Backend]({{page page='audit'}}#audit-back-ends)
```

Author: I'm updating the title on audit.md _Audit Back-ends_ to _Audit Backends_, but preserve the anchor `audit-back-ends` for backward.

based on a valid NXQL query whose `FROM` clause states the source backend.

> Constraints can be expressed on the `LogEntry` fields (`event_id`, `log_date`,

Reviewer: Is it `event_id` or `eventId`? Same for the other fields.

Author: It is `eventId`, etc... those might be taken from `LogEntryConstants` class.

backends — a second copy from the same source to the same target is rejected
as long as a previous one is still running.

Default action configuration:

| Property                                           | Default |
|----------------------------------------------------|---------|
| `nuxeo.bulk.action.copyAudit.defaultConcurrency`   | `2`     |
| `nuxeo.bulk.action.copyAudit.defaultPartitions`    | `4`     |

The action also tolerates failures: it retries up to 3 times with an
exponential back-off (`500ms` → `10s`) before giving up.

## Management REST API

Three Management endpoints expose the copy / verification flow:

| Endpoint                                | Purpose                                                     |
|-----------------------------------------|-------------------------------------------------------------|
| `POST /management/audit/copy`           | Trigger a `copyAudit` bulk action between two backends.     |
| `GET  /management/audit/checkSearch`    | Run the same NXQL on several backends and compare results.  |
| `GET  /management/audit/introspection`  | Get a PlantUML view of the live audit routing.              |

## Typical Blue/Green Migration

> 1. Deploy a secondary Audit Backend (the `green` backend, the live one being

Reviewer 2: ```suggestion
1. Deploy a secondary Audit Backend (referred to as the green backend, the live one being the default)
```

   the `default` backend) — see the [Audit Router]({{page page='audit-router'}})
   page.

> 2. Contribute a route to start mirroring live ingestion to the `green`

Reviewer: Maybe link to the Audit Router page?

Author: I'm adding a link to _route_ word to the Audit Router page.

   backend.
3. Trigger `POST /management/audit/copy?from=default&to=green` to copy the
   historical entries.

> 4. Monitor the bulk status using the [Bulk Endpoint]({{page space='rest-api' version='1' page='bulk-endpoint'}}).

Reviewer: ```suggestion
4. Monitor the bulk command status using the [Bulk Endpoint]({{page space='rest-api' version='1' page='bulk-endpoint'}}).
```

5. Validate using `GET /management/audit/checkSearch?backend=default&backend=green`.
6. Promote the `green` backend as the new default — this is a repackaging
   step that registers the `green` backend as the `default` backend and the

>    previous default as the secondary one.

Reviewer: Maybe add an example of (pointing to) how to do this? Though it's probably described elsewhere, when landing directly on this page, might be useful to have a reference.

Author: We discuss it with @bdelbosc, this is clearly missing in the current documentation, but as it is currently verbose (see confluence page), we would wait for ticket stated [here](https://github.com/nuxeo/doc.nuxeo.com-content/pull/2715#issuecomment-4418914323) and then document.

When NXP will be addressed we would add that somewhere, probably move/group the search blue/green documentation to have the global how to in one place.

>    previous default as the secondary one.

Reviewer 2: After a promotion you would more naturally retire (or delete) the previous default, not keep it as a secondary.

>    previous default as the secondary one.

Reviewer 2: ```suggestion
   previous default configuration can be removed.
```

>    Confluence page for the detailed procedure.

Author: Please delete _See the [Archi..._ sentence, Confluence is not publicly available.

---
title: Copy an Audit Backend
description: Copy log entries from one Audit Backend to another using the audit scroll, the copyAudit bulk action and the Management REST API.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - blue-green-migration
toc: true
tree_item_index: 780
---

{{#> callout type='warning' heading='Available since 2025.19'}}

> The audit scroller, the `copyAudit` bulk action and the related Management
> REST API endpoints are introduced in Nuxeo 2025.19 (epic NXENG-240, Blue/Green
> Audit migration).
> {{/callout}}

Author: ```suggestion
{{#> callout type='warning' heading='Available since 2025.19'}}
The audit scroller, the `copyAudit` bulk action and the related Management
REST API endpoints are introduced in Nuxeo 2025.19.
{{/callout}}
```

one is still running.

Default action configuration:

| Property                                           | Default |
|----------------------------------------------------|---------|
| `nuxeo.bulk.action.copyAudit.defaultConcurrency`   | `2`     |
| `nuxeo.bulk.action.copyAudit.defaultPartitions`    | `4`     |

The action also tolerates failures: it retries up to 3 times with an
exponential back-off (`500ms` → `10s`) before giving up.

## Management REST API

Three Management endpoints expose the copy / verification flow:

| Endpoint                                | Purpose                                                     |
|-----------------------------------------|-------------------------------------------------------------|
| `POST /management/audit/copy`           | Trigger a `copyAudit` bulk action between two backends.     |
| `GET  /management/audit/checkSearch`    | Run the same NXQL on several backends and compare results.  |
| `GET  /management/audit/introspection`  | Get a PlantUML view of the live audit routing.              |

The full reference, request/response samples and curl examples are documented
in the REST API reference: see

> [Audit Endpoint](https://doc.nuxeo.com/rest-api/1/audit-endpoint/).

Author: I think having a link in the _Learn More_ section is enough, remove this paragraph.

## Typical Blue/Green Migration

1. Deploy a secondary Audit Backend (`backend B`) — see the
   [Audit Router]({{page page='audit-router'}}) page.

> 2. (Optional) Contribute a route to start mirroring live ingestion to `backend B`.

Author: ```suggestion
2. Contribute a route to start mirroring live ingestion to `backend B`.
```
This step is not really Optional if your desire is to do a blue/green migration, without it you'll miss live events in the backend B.

3. Trigger `POST /management/audit/copy?from=default&to=backendB` to copy the
   historical entries.
4. Monitor the bulk status using the [Bulk Endpoint](https://doc.nuxeo.com/rest-api/1/bulk-endpoint/).
5. Validate using `GET /management/audit/checkSearch?backend=default&backend=backendB`.
6. Switch the `nuxeo.audit.backend.default.factory` to the new implementation

>    and decommission `backend A`.

Author: I think I will rephrase this one because it looks like we just need to change a property whereas this is not the case (have a look to the confluence page explaining that). But I still want something light for this point, try to find a sentence that states this action item.

Author: Also I think decommission backend A is here to mean delete the original backend? I would rather have an additional optional item to talk about that.

{{#> callout type='info' heading='Coming next — Purge Audit Migration'}}
The same building blocks (audit scroll + `copyAudit` style bulk action) will be
reused to provide an uninterrupted **Audit Purge** capability. This is tracked
by [NXENG-241](https://hyland.atlassian.net/browse/NXENG-241).

> {{/callout}}

Author: Please remove the jira ticket, it is not publicly accessible.
